presto-hive/src/main/java/com/facebook/presto/hive/HiveMaterializedViewUtils.java

```diff
@@ -170,66 +186,124 @@ public static MaterializedDataPredicates getMaterializedDataPredicates(
                 .collect(toImmutableList()));
     }
 
+    public static Optional<Map<String, String>> viewToBaseTableNonNullablePartitions(
+            ConnectorMaterializedViewDefinition viewDefinition,
+            Table baseTable)
+    {
+        SchemaTableName schemaBaseTable = new SchemaTableName(baseTable.getDatabaseName(), baseTable.getTableName());
+        if (!viewDefinition.getBaseTablesOnOuterJoinSide().contains(schemaBaseTable)) {
+            return Optional.empty();
+        }
+        Map<String, String> viewToBaseNonNullColumns = new HashMap<>();
+
+        Map<String, Map<SchemaTableName, String>> columnMappings = viewDefinition.getColumnMappingsAsMap();
+        Map<String, Map<SchemaTableName, String>> nonNullColumnMappings = viewDefinition.getNonNullColumnMappingsAsMap();
+
+        for (String viewPartition : viewDefinition.getValidRefreshColumns().orElse(ImmutableList.of())) {
+            String baseTablePartition = columnMappings.get(viewPartition).get(schemaBaseTable);
+            // Check if it is a base table partition
+            if (!baseTable.getPartitionColumns().stream().anyMatch(col -> col.getName().equals(baseTablePartition))) {
+                continue;
+            }
+            // Check if view partition maps to partition of base table, and to some partition of other base table where non-nullable mapping is present.
+            // For e.g. in case of left outer join, we want to find partition which maps to left table and we know that partition is non-nullable, unlike
+            // partition from right table which can become null if join conditions are not met.
+            if (nonNullColumnMappings.get(viewPartition).keySet().stream().allMatch(e -> !e.equals(schemaBaseTable)) &&
+                    nonNullColumnMappings.get(viewPartition).keySet().stream().anyMatch(t -> !viewDefinition.getBaseTablesOnOuterJoinSide().contains(t))) {
+                viewToBaseNonNullColumns.put(viewPartition, baseTablePartition);
+            }
+        }
+        return Optional.of(viewToBaseNonNullColumns);
+    }
+
+    public static MaterializedDataPredicates differenceDataPredicates(
+            MaterializedDataPredicates baseTablePredicatesInfo,
+            MaterializedDataPredicates viewPredicatesInfo,
+            Map<String, String> viewToBaseTablePredicatesKeyMap)
+    {
+        return differenceDataPredicates(baseTablePredicatesInfo, viewPredicatesInfo, viewToBaseTablePredicatesKeyMap, ImmutableMap.of());
+    }
+
     public static MaterializedDataPredicates differenceDataPredicates(
```

Comment:
While you are here, could you please add a function comment for this function with explanation on the input parameters and return result?